cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -420,6 +423,303 @@ class self_comparator {
+   *
+   * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+   */
+  class element_comparator {
+   public:
+    /**
+     * @brief Construct type-dispatched function object for comparing equality
+     * between two elements.
+     *
+     * @note `lhs` and `rhs` may be the same.
+     *
+     * @param has_nulls Indicates if either input column contains nulls.
+     * @param lhs The column containing the first element
+     * @param rhs The column containing the second element (may be the same as lhs)
+     * @param nulls_are_equal Indicates if two null elements are treated as equivalent
+     */
+    __device__ element_comparator(Nullate has_nulls,
+                                  column_device_view lhs,
+                                  column_device_view rhs,
+                                  null_equality nulls_are_equal = null_equality::EQUAL) noexcept
+      : lhs{lhs}, rhs{rhs}, nulls{has_nulls}, nulls_are_equal{nulls_are_equal}
+    {
+    }
+
+    /**
+     * @brief Compares the specified elements for equality.
+     *
+     * @param lhs_element_index The index of the first element
+     * @param rhs_element_index The index of the second element
+     * @return True if lhs and rhs are equal or if both lhs and rhs are null and nulls are
+     * configured to be considered equal (`nulls_are_equal` == `null_equality::EQUAL`)
+     */
+    template <typename Element, CUDF_ENABLE_IF(cudf::is_equality_comparable<Element, Element>())>
+    __device__ bool operator()(size_type const lhs_element_index,
+                               size_type const rhs_element_index) const noexcept
+    {
+      if (nulls) {
+        bool const lhs_is_null{lhs.is_null(lhs_element_index)};
+        bool const rhs_is_null{rhs.is_null(rhs_element_index)};
+        if (lhs_is_null and rhs_is_null) {
+          return nulls_are_equal == null_equality::EQUAL;
+        } else if (lhs_is_null != rhs_is_null) {
+          return false;
+        }
+      }
+
+      return equality_compare(lhs.element<Element>(lhs_element_index),
+                              rhs.element<Element>(rhs_element_index));
+    }
+
+    template <typename Element,
+              CUDF_ENABLE_IF(not cudf::is_equality_comparable<Element, Element>() and
+                             not cudf::is_nested<Element>()),
+              typename... Args>
+    __device__ bool operator()(Args...)
+    {
+      CUDF_UNREACHABLE("Attempted to compare elements of uncomparable types.");
+    }
+
+    template <typename Element, CUDF_ENABLE_IF(cudf::is_nested<Element>())>
+    __device__ bool operator()(size_type const lhs_element_index,
+                               size_type const rhs_element_index) const noexcept
+    {
+      column_device_view lcol = lhs.slice(lhs_element_index, 1);
+      column_device_view rcol = rhs.slice(rhs_element_index, 1);
+      while (is_nested(lcol.type())) {
+        if (nulls) {
```

Comment:
The `nulls` parameter already dispatched on nullability at compile time. As for nullability across levels, it can't be compile time but we could do something like this 
```c++
if (lcol.nullable() and rcol.nullable()) {
  auto lvalid = make_validity_iterator(lcol);
  auto rvalid = make_validity_iterator(rcol);
  if (nulls_are_equal == null_equality::UNEQUAL) {
    if (thrust::any_of(
          thrust::seq, lvalid, lvalid + lcol.size(), thrust::logical_not<bool>()) or
        thrust::any_of(
          thrust::seq, rvalid, rvalid + rcol.size(), thrust::logical_not<bool>())) {
      return false;
    }
  } else {
    if (not thrust::equal(thrust::seq, lvalid, lvalid + lcol.size(), rvalid)) {
      return false;
    }
  }
} else if (lcol.nullable()) {
  auto lvalid = make_validity_iterator(lcol);
  if (thrust::any_of(
        thrust::seq, lvalid, lvalid + lcol.size(), thrust::logical_not<bool>())) {
    return false;
  }
} else if (rcol.nullable()) {
  auto rvalid = make_validity_iterator(rcol);
  if (thrust::any_of(
        thrust::seq, rvalid, rvalid + rcol.size(), thrust::logical_not<bool>())) {
    return false;
  }
}
```

Reply:
It could be shortened with a lambda
```c++
if (nulls) {
  auto any_nulls_in = [](column_device_view const& c) {
    auto lvalid = make_validity_iterator(lcol);
    return thrust::any_of(
      thrust::seq, lvalid, lvalid + lcol.size(), thrust::logical_not<bool>());
  };
  if (lcol.nullable() and rcol.nullable()) {
    if (nulls_are_equal == null_equality::UNEQUAL) {
      if (any_nulls_in(lcol) or any_nulls_in(rcol)) { return false; }
    } else {
      auto lvalid = make_validity_iterator(lcol);
      auto rvalid = make_validity_iterator(rcol);
      if (not thrust::equal(thrust::seq, lvalid, lvalid + lcol.size(), rvalid)) {
        return false;
      }
    }
  } else if (lcol.nullable()) {
    if (any_nulls_in(lcol)) { return false; }
  } else if (rcol.nullable()) {
    if (any_nulls_in(rcol)) { return false; }
  }
}
```
We can remove the newly introduced "safe" validity iterator with this

Reply:
OK so we have a problem. Because we're only building `self_comparator` right now, we cannot test this. I'm stashing it for now and test+benchmark this when we add two table comparator.